src/Kvasir/Reconstitution/Lookups.cs

```diff
@@ -30,19 +30,30 @@ internal static partial class Lookup {
         public static object ByKey(DBData key, IEnumerable<object> entities, DataExtractionPlan extractor) {
             Debug.Assert(entities.All(e => e.GetType().IsInstanceOf(extractor.ExpectedSource)));
 
+            if (CacheDictionary.ContainsKey(key))
```

Comment:
I also think this is slightly pessimized: we have to do the lookup twice, once for the `ContainsKey` check and once for the `this[]` indexer. I think we can do this instead:
```
if (cacheDictionary_.TryGetValue(key, out object? result)) {
    return result!;
}
```

Might need to fiddle with the syntax a bit to deal with nullability concerns.

Reply:
I'm pretty sure `object`s are nullable by default, so `object?` is redundant. Moreover, Line 31's `Debug.Assert` throws an exception when an entity in `entities` is null - do we really have to worry about nullability? I agree that we could make the two lookups into one though!

Reply:
So . . . starting in C# 8.0, there's the notion of a "nullable reference type." It's all _technically_ just compiler sugar, but `T?` means "a T that can take a value of `null`". For value types (primitives, enumerations, and structs), `T?` is equivalent to `Nullable<T>` - that isn't changed. But now, the following will produce a compiler warning (or an error, depending on settings) if you are in a "nullable enabled context":
```
object obj = null;
```

The issue here is that `TryGetValue` has to take a default-initialized value for the return slot that it may not overwrite if the lookup fails. That's why my snippet has `object?` - it's saying "this is an `object`, and it _can_ be `null`" since it's default-initialized, which for reference types (like `object`) is `null`. The `!` operator in the return value tells the compiler "even if your static analysis says that this value might be `null`, trust me that it isn't." In this case, it's safe to do, because if the lookup succeeds (and we get into the body of the `if` statement), then `result` is guaranteed to be non-`null` because the underlying Dictionary has a non-`null` value type of `object`.

The reason that there's no check on Line 31 is that the type of the input argument is `IEnumerable<object>`, which means "a non-`null` instance of `IEnumerable<T>` where the elements are non-`null` instances of `object`. This is an internal-facing function, so I'm trusting myself to not pass a `null` argument or an argument that contains `null`s, hence the lack of an explicit check. So yes, that line _will_ throw a `NullReferenceException` if anything involved is `null`, but that's okay because it's (1) a debug-only check and (2) purely defensive - it should trigger in a unit test and result in a patch elsewhere.

Reply:
Okay, that makes sense!